File: src/ikc/main/Ioke.Lang/ControlFlow.cs

namespace Ioke.Lang {
    using System.Collections;

    public class ControlFlow : System.Exception {
        public class Exit : ControlFlow {
            int exitValue = 1;
            public Exit() : base(null) {}
            public Exit(object reason) : base(reason, "OH NO, exit out of place, because of: " + reason) {}
            public Exit(int value) : base(null) {
                this.exitValue = value;
            }
            public Exit(object reason, int value) : base(reason, "OH NO, exit out of place, because of: " + reason) {
                this.exitValue = value;
            }
            public int ExitValue {
                get { return exitValue; }
            }
        }

        public class Break : ControlFlow {
            public Break(object value) : base(value) {
            }
        }

        public class Continue : ControlFlow {
            public Continue() : base(null) {
            }
        }

        public class Return : ControlFlow {
            public readonly object context;
            public Return(object value, object context) : base(value) {
                this.context = context;
            }
        }

        public class Restart : ControlFlow {
            IList arguments;
            public Restart(Runtime.RestartInfo value, IList arguments) : base(value) {
                this.arguments = arguments;
            }

            public Runtime.RestartInfo GetRestart {
                get { return (Runtime.RestartInfo)Value; }
            }

            public IList Arguments {
                get { return this.arguments; }
            }
        }

        public class Rescue : ControlFlow {
            IokeObject condition;
            public Rescue(Runtime.RescueInfo value, IokeObject condition) : base(value) {
                this.condition = condition;
            }

            public Runtime.RescueInfo GetRescue {
                get { return (Runtime.RescueInfo)Value; }
            }

            public IokeObject Condition {
                get { return this.condition; }
            }

            public override string ToString() {
                return "rescue: " + Value.ToString();
            }
        }

        private object value;

        public ControlFlow(object value) {
            this.value = value;
        }

        public ControlFlow(object value, string message) : base(message) {
            this.value = value;
        }

        public object Value {
            get { return value; }
        }
    }
}

File: src/ikj/main/ioke/lang/exceptions/ControlFlow.java
/*
 * See LICENSE file in distribution for copyright and licensing information.
 */
package ioke.lang.exceptions;

import java.util.List;

import ioke.lang.Runtime;
import ioke.lang.IokeObject;

/**
 *
 * @author <a href="mailto:ola.bini@gmail.com">Ola Bini</a>
 */
public class ControlFlow extends Throwable {
    public static class Exit extends ControlFlow {
        private int exitValue = 1;
        public Exit() {
            super(null);
        }

        public Exit(Object reason) {
            super(reason, "OH NO, exit out of place, because of: " + reason);
        }

        public Exit(int value) {
            super(null);
            this.exitValue = value;
        }

        public Exit(Object reason, int value) {
            super(reason, "OH NO, exit out of place, because of: " + reason);
            this.exitValue = value;
        }

        public int getExitValue() {
            return exitValue;
        }
    }

    public static class Break extends ControlFlow {
        public Break(Object value) {
            super(value);
        }
    }

    public static class Continue extends ControlFlow {
        public Continue() {
            super(null);
        }
    }

    public static class Return extends ControlFlow {
        public final Object context;
        public Return(Object value, Object context) {
            super(value);
            this.context = context;
        }
    }

    public static class Restart extends ControlFlow {
        private List<Object> arguments;
        public Restart(Runtime.RestartInfo value, List<Object> arguments) {
            super(value);
            this.arguments = arguments;
        }

        public Runtime.RestartInfo getRestart() {
            return (Runtime.RestartInfo)getValue();
        }

        public List<Object> getArguments() {
            return this.arguments;
        }
    }

    public static class Rescue extends ControlFlow {
        private IokeObject condition;
        public Rescue(Runtime.RescueInfo value, IokeObject condition) {
            super(value);
            this.condition = condition;
        }

        public Runtime.RescueInfo getRescue() {
            return (Runtime.RescueInfo)getValue();
        }

        public IokeObject getCondition() {
            return this.condition;
        }

        public String toString() {
            return "rescue: " + getValue().toString();
        }
    }

    private Object value;

    public ControlFlow(Object value) {
        this.value = value;
    }

    public ControlFlow(Object value, String message) {
        super(message);
        this.value = value;
    }

    public Object getValue() {
        return value;
    }

    @Override
    public Throwable fillInStackTrace() {
        // we don't need any stack trace
        return this;
    }
}// ControlFlow
